### PR TITLE
Implement `io.readResource`

### PR DESCRIPTION
**Pull request by armanbilge**
Addresses #2503.

### CODE REVIEW

File: io/jvm/src/main/scala/fs2/io/ioplatform.scala [hunk 1]
```diff
@@ -137,4 +137,15 @@ private[fs2] trait ioplatform {
   def stdinUtf8[F[_]: Sync](bufSize: Int): Stream[F, String] =
     stdin(bufSize).through(text.utf8.decode)
 
+  /** Stream of bytes read asynchronously from the specified classloader resource. */
+  def readResource[F[_]](
+      name: String,
+      chunkSize: Int,
+      classLoader: ClassLoader = getClass().getClassLoader()
+  )(implicit F: Sync[F]): Stream[F, Byte] =
+    Stream.eval(F.delay(Option(classLoader.getResourceAsStream(name)))).flatMap {
```
Comment (armanbilge) [inline io/jvm/src/main/scala/fs2/io/ioplatform.scala:146]: Maybe this should be blocking 🤔

**Comment by mpilquist** [comment]
It might be nice to do this in a way which can use `Files` if the resource is a `jar:file` URL. We don't currently allow it because we have no way to construct a `java.nio.file.Path` via a URI.

BTW, we should also note that the resource path is absolute, not relative to a class. E.g., `classOf[Foo].getResource("foo.properties")` vs `classOf[Foo].getClassLoader.getResource("a/b/c/foo.properties")`

**Comment by armanbilge** [comment]
> It might be nice to do this in a way which can use Files if the resource is a jar:file URL.

How would that work exactly? Do you mean use `Files` to implement this method? Or do you mean add a method to the `Files` API which works with a `URL` instead of `Path`?

For the former, I'm not sure if that's possible, since the file is embedded in the jar right? Seems like [`JavaURLConnection`](https://docs.oracle.com/javase/8/docs/api/java/net/JarURLConnection.html) would be how to interact with those.

**Comment by mpilquist** [comment]
Yeah okay, I forgot that there's no default `jar:` file system handler. I wanted a way to do this without going through `InputStream` but I guess that's not possible.

**Comment by armanbilge** [comment]
Ohh, no I think I got you. Maybe this is possible, using [`FileSystems.newFileSystem`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/FileSystems.html)? Would this actually make a difference?

**Comment by mpilquist** [comment]
Right, we'd have to call `newFileSystem` with a path to the JAR and then create a `java.nio.file.Path` from that file system for the resource in the JAR. And we'd likely still need a fallback to current implementation. Probably best to stick with what you've got -- the original fs2 `InputStream` integration seemed to have high overhead but I'm probably worried about an implementation that no longer exists.

File: io/jvm/src/main/scala/fs2/io/ioplatform.scala [hunk 2]
```diff
@@ -137,4 +138,29 @@ private[fs2] trait ioplatform {
   def stdinUtf8[F[_]: Sync](bufSize: Int): Stream[F, String] =
     stdin(bufSize).through(text.utf8.decode)
 
+  /** Stream of bytes read asynchronously from the specified resource relative to the class `C`.
+    * @see [[readClassLoaderResource]] for a resource relative to a classloader.
+    */
+  def readClassResource[F[_], C](
+      name: String,
+      chunkSize: Int
+  )(implicit F: Sync[F], ct: ClassTag[C]): Stream[F, Byte] =
```
Comment (armanbilge) [inline io/jvm/src/main/scala/fs2/io/ioplatform.scala:147]: H/t @Daenyth for the classtag trick.

File: io/jvm/src/main/scala/fs2/io/ioplatform.scala [hunk 3]
```diff
@@ -137,4 +138,29 @@ private[fs2] trait ioplatform {
...
+      case Some(resource) => io.readInputStream(resource.pure, chunkSize)
+      case None           => Stream.raiseError(new IOException(s"Resource $name not found"))
+    }
+
+  /** Stream of bytes read asynchronously from the specified classloader resource.
+    * @see [[readClassResource]] for a resource relative to a class.
+    */
+  def readClassLoaderResource[F[_]](
+      name: String,
+      chunkSize: Int,
```
Comment (mpilquist) [inline io/jvm/src/main/scala/fs2/io/ioplatform.scala:158]: Default chunk size here and above to 64K like in Files?

Comment (armanbilge) [reply io/jvm/src/main/scala/fs2/io/ioplatform.scala:158]:
```suggestion
      chunkSize: Int = 64 * 1024,
```

File: io/jvm/src/main/scala/fs2/io/ioplatform.scala [hunk 4]
```diff
@@ -137,4 +138,29 @@ private[fs2] trait ioplatform {
   def stdinUtf8[F[_]: Sync](bufSize: Int): Stream[F, String] =
     stdin(bufSize).through(text.utf8.decode)
 
+  /** Stream of bytes read asynchronously from the specified resource relative to the class `C`.
+    * @see [[readClassLoaderResource]] for a resource relative to a classloader.
+    */
+  def readClassResource[F[_], C](
+      name: String,
+      chunkSize: Int
```
Comment (armanbilge) [inline io/jvm/src/main/scala/fs2/io/ioplatform.scala:146]:
```suggestion
      chunkSize: Int = 64 * 1024
```

**Pull request merged**